esphome/components/json/json_util.cpp

```diff
@@ -23,21 +23,21 @@ std::string build_json(const json_build_t &f) {
 #ifdef USE_ESP8266
   const size_t free_heap = ESP.getMaxFreeBlockSize();  // NOLINT(readability-static-accessed-through-instance)
 #elif defined(USE_ESP32)
-  const size_t free_heap = heap_caps_get_largest_free_block(MALLOC_CAP_8BIT);
+  const size_t free_heap = heap_caps_get_largest_free_block(MALLOC_CAP_INTERNAL);
 #endif
 
-  const size_t request_size = std::min(free_heap, (size_t) 512);
+  const size_t request_size = std::min(free_heap - 2048, (size_t) 5120);
 
   DynamicJsonDocument json_document(request_size);
-  if (json_document.capacity() == 0) {
+  if (json_document.memoryPool().buffer() == nullptr) {
```

Comment:
```suggestion
    iif (json_document.capacity() == 0) {
```

We don't need to revert this